examples/pennylane_qubit_rotation.py

```diff
@@ -0,0 +1,367 @@
+# the derivative of the QNode with respect to the argument specified in ``argnum``.
+# In this case, the function ``circuit`` takes one argument (``params``), so we
+# specify ``argnum=0``. Because the argument has two elements, the returned gradient
+# is two-dimensional. We can then evaluate this gradient function at any point in the parameter space.
+
+print(dcircuit([0.54, 0.12]))
+
+################################################################################
+# **A note on arguments**
+#
+# Quantum circuit functions, being a restricted subset of Python functions,
+# can also make use of multiple positional arguments and keyword arguments.
+# For example, we could have defined the above quantum circuit function using
+# two positional arguments, instead of one array argument:
+
+
+@qml.qnode(dev1)
+def circuit2(phi1, phi2):
+    qml.RX(phi1, wires=0)
+    qml.RY(phi2, wires=0)
+    return qml.expval(qml.PauliZ(0))
+
+
+################################################################################
+# When we calculate the gradient for such a function, the usage of ``argnum``
+# will be slightly different. In this case, ``argnum=0`` will return the gradient
+# with respect to only the first parameter (``phi1``), and ``argnum=1`` will give
+# the gradient for ``phi2``. To get the gradient with respect to both parameters,
+# we can use ``argnum=[0,1]``:
+
+dcircuit = qml.grad(circuit2, argnum=[0, 1])
+print(dcircuit(0.54, 0.12))
+
+################################################################################
+# Keyword arguments may also be used in your custom quantum function. PennyLane
+# does **not** differentiate QNodes with respect to keyword arguments,
+# so they are useful for passing external data to your QNode.
+
+
+################################################################################
+# Optimization
+# ------------
+#
+# .. admonition:: Definition
+#     :class: defn
+#
+#     If using the default NumPy/Autograd interface, PennyLane provides a collection
+#     of optimizers based on gradient descent. These optimizers accept a cost function
+#     and initial parameters, and utilize PennyLane's automatic differentiation
+#     to perform gradient descent.
+#
+# .. tip::
+#
+#    *See* :mod:`pennylane.optimize` *for details and documentation of available optimizers*
+#
+# Next, let's make use of PennyLane's built-in optimizers to optimize the two circuit
+# parameters :math:`\phi_1` and :math:`\phi_2` such that the qubit, originally in state
+# :math:`\ket{0}`, is rotated to be in state :math:`\ket{1}`. This is equivalent to measuring a
+# Pauli-Z expectation value of :math:`-1`, since the state :math:`\ket{1}` is an eigenvector
+# of the Pauli-Z matrix with eigenvalue :math:`\lambda=-1`.
+#
+# In other words, the optimization procedure will find the weights
+# :math:`\phi_1` and :math:`\phi_2` that result in the following rotation on the Bloch sphere:
+#
+# .. figure:: ../../examples/figures/bloch.png
```

Comment:
Can we represent figures as html references, and link to the correct location on the github repo? Might let us render images properly in notebooks

Reply:
Perhaps relevant: https://sphinx-gallery.github.io/advanced.html#write-a-custom-image-scraper

Reply:
I think we tried it but I cant recall if it worked...

Reply:
Aroosa and I tried this, but the developmental workflow wasn't very optimal (it required a preliminary PR to get the URL of newly added images). The end result ended up looking poorer in the documentation, and not too great (although it rendered) in the notebook.